long-ride-alerts/src/main/java/org/apache/flink/training/exercises/longrides/LongRidesExercise.java

```diff
@@ -98,17 +102,62 @@ public static void main(String[] args) throws Exception {
     @VisibleForTesting
     public static class AlertFunction extends KeyedProcessFunction<Long, TaxiRide, Long> {
 
+        private static final long RIDE_DURATION_ALERT_TIME_IN_MS = 1000 * 60 * 60 * 2;
+        private ValueState<TaxiRide> startTaxiRide;
+        private ValueState<TaxiRide> endTaxiRide;
+
         @Override
-        public void open(Configuration config) throws Exception {
-            throw new MissingSolutionException();
+        public void open(Configuration config) {
+            startTaxiRide = getRuntimeContext().getState(new ValueStateDescriptor<>("startTaxiRide", TaxiRide.class));
+            endTaxiRide = getRuntimeContext().getState(new ValueStateDescriptor<>("endTaxiRide", TaxiRide.class));
         }
 
         @Override
         public void processElement(TaxiRide ride, Context context, Collector<Long> out)
-                throws Exception {}
+                throws Exception {
+            if (ride.isStart) {
+                TaxiRide taxiRide = endTaxiRide.value();
+                if (taxiRide != null) {
+                    handleRideDuration(ride, taxiRide, context, out);
+                    endTaxiRide.clear();
+                } else {
+                    startTaxiRide.update(ride);
```

Comment:
Is state cleared in all situations?
From the readme:
> An END event may be missing, but you may assume there are no duplicated events, and no missing START events.
A start event arrives, you save the startTaxiRide state, and the timer triggers (because the end event is missing) and triggers the alert, but leaving the start event in the state.

Reply:
cleared the start state in the timer, thanks!